File: backend_python/app/search/espacenet.py
"""Espacenet patent-database search provider.

Free tier: 20K requests per month, no API key required for basic queries.

Note: this is a faithful port of the Go implementation, which targets
the legacy `cgi-bin/espacenet` endpoint and may not return JSON in
production. Failures fall through gracefully via `multi_search`.
"""

from __future__ import annotations

import logging
from datetime import datetime
from typing import List
from urllib.parse import quote_plus

import httpx

from app.search.base import SearchProvider, SearchQuery, SearchResult
from app.search.registry import register_provider

logger = logging.getLogger(__name__)


def _build_espacenet_query(keywords: List[str]) -> str:
    keywords = [k.strip() for k in keywords if k.strip()][:5]
    return " AND ".join(keywords)


class EspacenetProvider(SearchProvider):
    BASE_URL = "https://www.espacenet.com/cgi-bin/espacenet"

    def name(self) -> str:
        return "espacenet"

    def _pick_keywords(self, query: SearchQuery) -> List[str]:
        for lang in ("EN", "DE", "ZH"):
            kws = query.keywords.get(lang) or []
            if kws:
                return kws
        return []

    async def search(self, query: SearchQuery) -> List[SearchResult]:
        keywords = self._pick_keywords(query)
        if not keywords:
            raise RuntimeError("no keywords provided for Espacenet search")

        search_query = _build_espacenet_query(keywords)

        url = (
            f"{self.BASE_URL}"
            f"?action=Search&CL=&QUERY={quote_plus(search_query)}"
            f"&STR=&DB=espacenet&FIRST=1&NUM={query.max_results}&format=json"
        )

        headers = {"User-Agent": "Mozilla/5.0 (compatible; Metheriel/1.0)"}

        async with httpx.AsyncClient(timeout=15.0, headers=headers) as client:
            resp = await client.get(url)

        if resp.status_code != 200:
            raise RuntimeError(
                f"espacenet API error: status {resp.status_code} - {resp.text[:200]}"
            )

        try:
            payload = resp.json()
        except ValueError as exc:
            raise RuntimeError(f"failed to parse Espacenet response: {exc}") from exc

        entries = payload.get("results") or []
        results: List[SearchResult] = []
        for entry in entries:
            title = (entry.get("title") or "").strip()
            if not title:
                continue

            pub_date_str = entry.get("publication_date") or ""
            pub_date: datetime = datetime.utcnow()
            if pub_date_str:
                try:
                    pub_date = datetime.strptime(pub_date_str, "%Y-%m-%d")
                except ValueError:
                    pass

            if query.target_date is not None and pub_date.replace(tzinfo=None) > query.target_date.replace(tzinfo=None):
                continue

            abstract = (
                entry.get("abstract_en")
                or entry.get("abstract_de")
                or entry.get("abstract_fr")
                or entry.get("applicant")
                or ""
            ).strip()

            patent_id = entry.get("patent_id") or ""
            results.append(
                SearchResult(
                    title=title,
                    url=f"https://www.espacenet.com/patent/ES/{patent_id}",
                    snippet=abstract,
                    date=pub_date,
                    source="Espacenet",
                    language="EN",
                )
            )

        return results


register_provider(EspacenetProvider())

File: backend/internal/search/espacenet.go
package search

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// EspacenetProvider searches the Espacenet patent database
// Free tier: 20K requests per month, no API key required for basic queries
type EspacenetProvider struct {
	baseURL string
	client  *http.Client
}

func NewEspacenetProvider() *EspacenetProvider {
	return &EspacenetProvider{
		baseURL: "https://www.espacenet.com/cgi-bin/espacenet",
		client: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

func (e *EspacenetProvider) Name() string {
	return "espacenet"
}

// buildEspacenetQuery constructs a query string for Espacenet search
// Espacenet accepts keyword queries and returns patent results
func buildEspacenetQuery(keywords []string) string {
	// Use AND operator for core keywords, max 5 to avoid timeouts
	maxKeywords := 5
	if len(keywords) > maxKeywords {
		keywords = keywords[:maxKeywords]
	}

	var parts []string
	for _, kw := range keywords {
		kw = strings.TrimSpace(kw)
		if kw == "" {
			continue
		}
		parts = append(parts, kw)
	}

	// Join with AND for strict matching
	return strings.Join(parts, " AND ")
}

// espacenetResult represents a single patent result from Espacenet JSON response
type espacenetResult struct {
	Title       string `json:"title"`
	AbstractEn  string `json:"abstract_en"`
	AbstractDe  string `json:"abstract_de"`
	AbstractFr  string `json:"abstract_fr"`
	PatentID    string `json:"patent_id"`
	PublicationDate string `json:"publication_date"`
	Applicant   string `json:"applicant"`
}

// espacenetResponse represents the JSON response structure from Espacenet
type espacenetResponse struct {
	Results []espacenetResult `json:"results"`
	Count   int               `json:"count"`
}

// Search queries Espacenet for patents matching the search query
func (e *EspacenetProvider) Search(
	ctx context.Context,
	query SearchQuery,
) ([]SearchResult, error) {

	// Get English keywords (fallback to other languages if needed)
	var keywords []string
	if enKeywords, ok := query.Keywords["EN"]; ok && len(enKeywords) > 0 {
		keywords = enKeywords
	} else if deKeywords, ok := query.Keywords["DE"]; ok && len(deKeywords) > 0 {
		keywords = deKeywords
	} else if zhKeywords, ok := query.Keywords["ZH"]; ok && len(zhKeywords) > 0 {
		keywords = zhKeywords
	}

	if len(keywords) == 0 {
		return nil, fmt.Errorf("no keywords provided for Espacenet search")
	}

	searchQuery := buildEspacenetQuery(keywords)

	// Construct Espacenet search URL
	// Using their free JSON  endpoint
	reqURL := fmt.Sprintf(
		"%s?action=Search&CL=&QUERY=%s&STR=&DB=espacenet&FIRST=%d&NUM=%d&format=json",
		e.baseURL,
		url.QueryEscape(searchQuery),
		1,
		query.MaxResults,
	)

	req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
	if err != nil {
		return nil, err
	}

	// Set User-Agent to avoid blocking
	req.Header.Set("User-Agent", "Mozilla/5.0 (compatible; Metheriel/1.0)")

	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("espacenet request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("espacenet API error: status %d - %s", resp.StatusCode, string(body))
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// Parse JSON response
	var espacenetResp espacenetResponse
	if err := json.Unmarshal(body, &espacenetResp); err != nil {
		// If JSON parsing fails, Espacenet might be returning HTML - return error
		return nil, fmt.Errorf("failed to parse Espacenet response: %w", err)
	}

	// Convert to SearchResult format
	var results []SearchResult
	for _, entry := range espacenetResp.Results {
		if entry.Title == "" {
			continue
		}

		// Parse publication date
		pubDate := time.Now()
		if entry.PublicationDate != "" {
			if parsed, err := time.Parse("2006-01-02", entry.PublicationDate); err == nil {
				pubDate = parsed
			}
		}

		// Filter by target date if specified
		if !query.TargetDate.IsZero() && pubDate.After(query.TargetDate) {
			continue
		}

		// Prefer English abstract, fallback to German or French
		abstract := entry.AbstractEn
		if abstract == "" {
			abstract = entry.AbstractDe
		}
		if abstract == "" {
			abstract = entry.AbstractFr
		}
		if abstract == "" {
			abstract = entry.Applicant // Fallback to applicant info
		}

		result := SearchResult{
			Title:    strings.TrimSpace(entry.Title),
			URL:      fmt.Sprintf("https://www.espacenet.com/patent/ES/%s", entry.PatentID),
			Snippet:  strings.TrimSpace(abstract),
			Date:     pubDate,
			Source:   "Espacenet",
			Language: "EN",
		}

		results = append(results, result)
	}

	return results, nil
}

// init registers the Espacenet provider
func init() {
	RegisterProvider(NewEspacenetProvider())
}
